storage/gc_queue.go

```diff
@@ -247,52 +257,59 @@ func (gcq *gcQueue) timer() time.Duration {
-// transaction and resolve the intent. If the transaction cannot be
-// aborted or intent cannot be resolved, the oldestIntentNanos value
-// is atomically updated to the min of oldestIntentNanos and the
-// intent's timestamp. The wait group is signaled on completion.
-func (gcq *gcQueue) resolveIntent(rng *Replica, key proto.Key, meta *engine.MVCCMetadata,
+// resolveIntents resolves the slice of intents (all gated by the same
+// transaction) by attempting to abort the txn via push. If the
+// transaction cannot be aborted or intent cannot be resolved, the
+// oldestIntentNanos value is atomically updated to the min of
+// oldestIntentNanos and the intent's timestamp. The wait group is
+// signaled on completion.
+func (gcq *gcQueue) resolveIntents(rng *Replica, txn proto.Transaction, intentKeys []proto.Key,
 	updateOldestIntent func(int64), wg *sync.WaitGroup) {
 	defer wg.Done() // signal wait group always on completion
 
-	log.Infof("resolving intent at %q ts=%s", key, meta.Timestamp)
+	if log.V(1) {
+		log.Infof("resolving intents for txn %s ts=%s", txn, txn.OrigTimestamp)
+	}
 
 	// Attempt to push the transaction which created the intent.
 	now := rng.rm.Clock().Now()
 	pushArgs := &proto.PushTxnRequest{
 		RequestHeader: proto.RequestHeader{
 			Timestamp:    now,
-			Key:          meta.Txn.Key,
+			Key:          txn.Key,
 			User:         security.RootUser,
 			UserPriority: gogoproto.Int32(proto.MaxPriority),
 			Txn:          nil,
 		},
 		Now:       now,
-		PusheeTxn: *meta.Txn,
+		PusheeTxn: txn,
 		PushType:  proto.ABORT_TXN,
 	}
 	pushReply := &proto.PushTxnResponse{}
 	b := &client.Batch{}
 	b.InternalAddCall(proto.Call{Args: pushArgs, Reply: pushReply})
 	if err := rng.rm.DB().Run(b); err != nil {
-		log.Warningf("push of txn %s failed: %s", meta.Txn, err)
-		updateOldestIntent(meta.Timestamp.WallTime)
+		log.Warningf("push of txn %s failed: %s", txn, err)
+		updateOldestIntent(txn.OrigTimestamp.WallTime)
 		return
 	}
 
-	// We pushed the transaction successfully, so resolve the intent.
-	resolveArgs := &proto.ResolveIntentRequest{
-		RequestHeader: proto.RequestHeader{
-			Timestamp: now,
-			Key:       key,
-			User:      security.RootUser,
-			Txn:       pushReply.PusheeTxn,
-		},
-	}
-	if _, err := rng.AddCmd(rng.context(), resolveArgs); err != nil {
-		log.Warningf("resolve of key %q failed: %s", key, err)
-		updateOldestIntent(meta.Timestamp.WallTime)
+	// We pushed the transaction successfully, so resolve the intents.
+	// TODO(spencer): use a batch here when available.
+	for _, key := range intentKeys {
+		resolveArgs := &proto.ResolveIntentRequest{
+			RequestHeader: proto.RequestHeader{
+				Timestamp: now,
+				Key:       key,
+				User:      security.RootUser,
+				Txn:       pushReply.PusheeTxn,
+			},
+		}
+		if _, err := rng.AddCmd(rng.context(), resolveArgs); err != nil {
+			log.Warningf("resolve of key %q failed: %s", key, err)
+			updateOldestIntent(txn.OrigTimestamp.WallTime)
+			return
```

Comment:
Right, but the goroutine can run a trivial wrapper fn that does the logging. I don't feel strongly